contrib/http_base/src/main/java/io/opentelemetry/contrib/http/core/AbstractHttpHandler.java

```diff
@@ -0,0 +1,204 @@
+   * @param compressedMessageSize size of the message after compressed (optional).
+   */
+  static void recordMessageEvent(
+      Span span,
+      long id,
+      AttributeValue type,
+      long uncompressedMessageSize,
+      long compressedMessageSize) {
+    Map<String, AttributeValue> attributes = new HashMap<>();
+    attributes.put(HttpTraceConstants.MSG_EVENT_ATTR_TYPE, type);
+    attributes.put(HttpTraceConstants.MSG_EVENT_ATTR_ID, AttributeValue.longAttributeValue(id));
+    if (compressedMessageSize > 0L) {
+      attributes.put(
+          HttpTraceConstants.MSG_EVENT_ATTR_COMPRESSED,
+          AttributeValue.longAttributeValue(compressedMessageSize));
+    }
+    if (uncompressedMessageSize > 0L) {
+      attributes.put(
+          HttpTraceConstants.MSG_EVENT_ATTR_UNCOMPRESSED,
+          AttributeValue.longAttributeValue(uncompressedMessageSize));
+    }
+    span.addEvent(HttpTraceConstants.MSG_EVENT_NAME, attributes);
+  }
+
+  private static void putAttributeIfNotEmptyOrNull(Span span, String key, @Nullable String value) {
+    if (value != null && !value.isEmpty()) {
+      span.setAttribute(key, AttributeValue.stringAttributeValue(value));
+    }
+  }
+
+  /**
+   * Instrument an HTTP span after a message is sent. Typically called for every chunk of request or
+   * response is sent.
+   *
+   * @param context request specific {@link HttpRequestContext}
+   * @param bytes bytes sent.
+   */
+  public final void handleMessageSent(HttpRequestContext context, long bytes) {
+    checkNotNull(context, "context is required");
+    context.sentMessageSize.addAndGet(bytes);
+    recordMessageEvent(
+        context.span,
+        context.sentSeqId.addAndGet(1L),
+        HttpTraceConstants.MSG_EVENT_ATTR_SENT,
+        bytes,
+        0L);
+  }
+
+  /**
+   * Instrument an HTTP span after a message is received. Typically called for every chunk of
+   * request or response is received.
+   *
+   * @param context request specific {@link HttpRequestContext}
+   * @param bytes bytes received.
+   */
+  public final void handleMessageReceived(HttpRequestContext context, long bytes) {
+    checkNotNull(context, "context is required");
+    context.receiveMessageSize.addAndGet(bytes);
+    recordMessageEvent(
+        context.span,
+        context.receviedSeqId.addAndGet(1L),
+        HttpTraceConstants.MSG_EVENT_ATTR_RECEIVED,
+        bytes,
+        0L);
+  }
+
+  void spanEnd(Span span, int httpStatus, @Nullable Throwable error) {
+    span.setAttribute(
+        HttpTraceConstants.HTTP_STATUS_CODE, AttributeValue.longAttributeValue(httpStatus));
+    if (error != null) {
+      String message = extractErrorMessage(error);
+      span.setAttribute(
+          HttpTraceConstants.HTTP_STATUS_TEXT, AttributeValue.stringAttributeValue(message));
+      getLogger().log(Level.FINE, message, error);
+    }
+    span.setStatus(statusConverter.convert(httpStatus));
+    span.end();
+  }
+
+  final String getSpanName(Q request) {
```

Comment:
this needs javadoc to explain how the span name is getting generated. Also, can we change it to `extractSpanName`? Is there a reason this needs to be final? Couldn't a specific instrumentation want to name spans differently?

Reply:
Fixed